File: lib-extension/src/main/java/com/xiaoyv/flexiflix/extension/java/model/MediaSourceInfo.kt
package com.xiaoyv.flexiflix.extension.java.model

import android.os.Parcelable
import kotlinx.parcelize.Parcelize

/**
 * [MediaSourceInfo] map to [com.xiaoyv.flexiflix.extension.annotation.MediaSource]
 *
 * @author why
 * @since 5/8/24
 */
@Parcelize
data class MediaSourceInfo(

    /**
     * 数据源唯一ID
     */
    val id: String,

    /**
     * 数据源名称
     */
    val name: String,

    /**
     * 数据源描述
     */
    val description: String,

    /**
     * 数据源作者
     */
    val author: String,

    /**
     * 成人内容
     */
    val nsfw: Boolean,

    /**
     * 数据源版本代码
     */
    val versionCode: Int,

    /**
     * 数据源版本名称
     */
    val versionName: String
) : Parcelable {
    companion object {
        @JvmStatic
        fun loadFromMap(infoMap: Map<String, String>): MediaSourceInfo {
            return MediaSourceInfo(
                id = infoMap.getOrDefault(
                    key = MediaSourceInfo::id.name,
                    defaultValue = ""
                ),
                name = infoMap.getOrDefault(
                    key = MediaSourceInfo::name.name,
                    defaultValue = "名称不见了"
                ),
                description = infoMap.getOrDefault(
                    key = MediaSourceInfo::description.name,
                    defaultValue = "暂时没有描述"
                ),
                author = infoMap.getOrDefault(
                    key = MediaSourceInfo::author.name,
                    defaultValue = "未知"
                ),
                nsfw = infoMap.getOrDefault(
                    key = MediaSourceInfo::nsfw.name,
                    defaultValue = "false"
                ).toBoolean(),
                versionCode = infoMap.getOrDefault(
                    key = MediaSourceInfo::versionCode.name,
                    defaultValue = "1"
                ).toIntOrNull() ?: 1,
                versionName = infoMap.getOrDefault(
                    key = MediaSourceInfo::versionName.name,
                    defaultValue = "v1.0.0"
                )
            )
        }
    }
}

File: web-nodejs/extension-common/src/model/MediaSourceInfo.ts
/**
 * MediaSourceInfo
 *
 * @author why
 * @since 5/8/24
 */
export interface MediaSourceInfo {
    /**
     * 插件数据源ID，需要全局唯一，建议搞得抽象一点
     *
     * 仅支持数字、字母、下划线、短横线组合。
     */
    id: string

    /**
     * 数据源名称
     */
    name: string

    /**
     * 数据源描述
     */
    description: string

    /**
     * 图标链接
     *
     * - 支持网络图片
     * - 支持 base64 编码的图片数据，固定格式为 http://localhost/api/image?data=[ImageBase64Data]，ImageBase64Data 为完整的图片 Base64 编码
     *
     * 示例链接：http://localhost/api/image?data=data:image/jpeg;base64,FASLKFKAMCZLFAKFQALCAS;FALFQPF==ASF=...
     */
    icon: string

    /**
     * 数据源作者
     */
    author: string

    /**
     * 成人内容
     */
    nsfw: boolean

    /**
     * 数据源版本代码
     */
    versionCode: number

    /**
     * 数据源版本名称
     */
    versionName: string
}
